src/src/rx-serial/devSerialIO.cpp

```diff
@@ -29,20 +36,131 @@ static int event()
+            if (newTeamracePosition != config.GetTeamracePosition())
+                teamraceOutputInhibitState = troiDisableAwaitConfirm;
+            break;
+
+        case troiDisableAwaitConfirm:
+            // Must receive the same new position twice in a row for state to change
+            if (lastTeamracePosition == newTeamracePosition)
+            {
+                if (newTeamracePosition != config.GetTeamracePosition())
+                    teamraceOutputInhibitState = troiInhibit; // disable output
+                else
+                    teamraceOutputInhibitState = troiPass; // return to normal
+            }
+            break;
+
+        case troiInhibit:
+            // User appears to be switching to this model, wait for confirm
+            if (newTeamracePosition == config.GetTeamracePosition())
+                teamraceOutputInhibitState = troiEnableAwaitConfirm;
+            break;
+
+        case troiEnableAwaitConfirm:
+            // Must receive the same new position twice in a row for state to change
+            if (lastTeamracePosition == newTeamracePosition)
+            {
+                if (newTeamracePosition == config.GetTeamracePosition())
+                    teamraceOutputInhibitState = troiPass; // return to normal
+                else
+                    teamraceOutputInhibitState = troiInhibit; // back to disabled
+            }
+            break;
+    }
+
+    lastTeamracePosition = newTeamracePosition;
+    // troiPass or troiDisablePending indicate the model is selected still,
+    // however returning true if troiDisablePending means this RX could send
+    // telemetry and we do not want that
+    teamraceHasModelMatch = teamraceOutputInhibitState == troiPass;
+    return retVal;
+}
+
 static int timeout()
 {
     if (connectionState == serialUpdate)
     {
         return DURATION_NEVER;  // stop callbacks when doing serial update
     }
 
-    uint32_t duration = 10; // 10ms callback (i.e. when no theres no model match)
-    // only send frames if we have a model match
-    if (connectionHasModelMatch)
-    {
-        duration = serialIO->sendRCFrame(frameAvailable, ChannelData);
-    }
-    frameAvailable = false;
+    /***
+     * TODO: This contains a problem!!
+     * confirmFrameAvailable() is designed to be the thing that determines if RC frames are to be sent out
+     * which includes:
+     *      - No data received to be sent (interpacket delay)
+     *      - Connection does not have model match
+     *      - TeamRace enabled but different position selected
+     * However, the SBUS IO writer doesn't go off new data coming in, it just sends data at a 9ms cadence
+     * and therefore does not respect any of these conditions, relying on the one-off "failsafe" member
+     * modelmatch was addressed in #2211, but resolving the merge conflict here (capnbry) re-breaks it
+     *
+     * Commiting this anyway though to work out a better resolution
+    */
```

Comment:
Does this bit still need addressing?

Reply:
Would #2457 help with this?

Reply:
The comment still is relevant. I sort of hacked in the needed change directly into the SBUS IO so it works with teamrace, but ideally all of that should be happening here rather than copying all the logic into the individual IO handlers.

The logic needs to unify the failsafe behavior (which becomes a little more disjointed via #2535 too and should be accounted for) and then remove the code from the beginning of SBUS and abstract it into the base. I haven't given it much thought yet because the failsafe changes are a key factor in how that gets handled, so the comment should stay until that's worked out so no relevant information is forgotten.